wlfreerdp crashing with "malloc(): unaligned tcache chunk detected"

**Found a bug? - We would like to help you and smash the bug away.**
1. If it's a __new__ bug - create a new issue.
1. For more details see https://github.com/FreeRDP/FreeRDP/wiki/BugReporting


**Describe the bug**
During a session, the wlfreerdp crashes. All I have is the malloc error:

```
...
[15:08:39:599] [19612:19612] [INFO][com.freerdp.client.wayland.cliprdr] - client announces mime SAVE_TARGETS
malloc(): unaligned tcache chunk detected
[15:08:41:624] [19612:19612] [INFO][com.freerdp.utils] - Caught signal 'Aborted' [6]

```

**To Reproduce**
Steps to reproduce the behavior:
1. Connect to windows server
2. Wait a couple of minutes

**Expected behavior**
It should not crash

**Application details**
* FreeRDP version 2.4.1
* Command line used: wlfreerdp /drive:user@domain,/home/domain/user/rdp +fonts /audio-mode:0 /cert-tofu /w:1240 /h:800 /ipv6 /gdi:hw /rfx /dynamic-resolution /d:domain /u:user /v:server /port:3389 -themes -wallpaper
* Output of :
```This is FreeRDP version 2.4.1 (n/a)
Build configuration: BUILD_TESTING=OFF BUILTIN_CHANNELS=OFF HAVE_AIO_H=1 HAVE_EXECINFO_H=1 HAVE_FCNTL_H=1 HAVE_GETLOGIN_R=1 HAVE_INTTYPES_H=1 HAVE_JOURNALD_H=TRUE HAVE_MATH_C99_LONG_DOUBLE=1 HAVE_POLL_H=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK=ON HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIBS= HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL=1 HAVE_SYSLOG_H=1 HAVE_SYS_EVENTFD_H=1 HAVE_SYS_FILIO_H= HAVE_SYS_MODEM_H= HAVE_SYS_SELECT_H=1 HAVE_SYS_SOCKIO_H= HAVE_SYS_STRTIO_H= HAVE_SYS_TIMERFD_H=1 HAVE_TM_GMTOFF=1 HAVE_UNISTD_H=1 HAVE_XI_TOUCH_CLASS=1 WITH_ALSA=ON WITH_CAIRO=ON WITH_CCACHE=ON WITH_CHANNELS=ON WITH_CLANG_FORMAT=ON WITH_CLIENT=ON WITH_CLIENT_AVAILABLE=1 WITH_CLIENT_CHANNELS=ON WITH_CLIENT_CHANNELS_AVAILABLE=1 WITH_CLIENT_COMMON=ON WITH_CLIENT_INTERFACE=OFF WITH_CUPS=ON WITH_DEBUG_ALL=OFF WITH_DEBUG_CAPABILITIES=OFF WITH_DEBUG_CERTIFICATE=OFF WITH_DEBUG_CHANNELS=OFF WITH_DEBUG_CLIPRDR=OFF WITH_DEBUG_DVC=OFF WITH_DEBUG_KBD=OFF WITH_DEBUG_LICENSE=OFF WITH_DEBUG_MUTEX=OFF WITH_DEBUG_NEGO=OFF WITH_DEBUG_NLA=OFF WITH_DEBUG_NTLM=OFF WITH_DEBUG_RAIL=OFF WITH_DEBUG_RDP=OFF WITH_DEBUG_RDPDR=OFF WITH_DEBUG_RDPEI=OFF WITH_DEBUG_RDPGFX=OFF WITH_DEBUG_REDIR=OFF WITH_DEBUG_RFX=OFF WITH_DEBUG_RINGBUFFER=OFF WITH_DEBUG_SCARD=OFF WITH_DEBUG_SND=OFF WITH_DEBUG_SVC=OFF WITH_DEBUG_SYMBOLS=OFF WITH_DEBUG_THREADS=OFF WITH_DEBUG_TIMEZONE=OFF WITH_DEBUG_TRANSPORT=OFF WITH_DEBUG_TSG=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF_AVAILABLE=0 WITH_DEBUG_URBDRC=OFF WITH_DEBUG_WND=OFF WITH_DEBUG_X11=OFF WITH_DEBUG_X11_CLIPRDR=OFF WITH_DEBUG_X11_LOCAL_MOVESIZE=OFF WITH_DEBUG_XV=OFF WITH_DIRECTFB=OFF WITH_DSP_EXPERIMENTAL=OFF WITH_DSP_FFMPEG=OFF WITH_EVENTFD_READ_WRITE=1 WITH_FAAC=OFF WITH_FAAD2=OFF WITH_FFMPEG=TRUE WITH_FFMPEG=TRUE WITH_GFX_H264=ON WITH_GPROF=OFF WITH_GSM=ON WITH_GSSAPI=OFF WITH_GSTREAMER_0_10=OFF WITH_GSTREAMER_1_0=ON WITH_ICU=ON WITH_IPP=OFF WITH_JPEG=ON WITH_LAME=ON WITH_LIBRARY_VERSIONING=ON WITH_LIBSYSTEMD=ON WITH_MACAUDIO=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO_AVAILABLE=0 WITH_MANPAGES=ON WITH_MBEDTLS=OFF WITH_OPENCL=OFF WITH_OPENH264=OFF WITH_OPENSLES=OFF WITH_OPENSSL=ON WITH_OSS=ON WITH_PAM=ON WITH_PCSC=ON WITH_PROFILER=OFF WITH_PROXY=ON WITH_PROXY_MODULES=OFF WITH_PULSE=ON WITH_SAMPLE=OFF WITH_SANITIZE_ADDRESS=OFF WITH_SANITIZE_ADDRESS_AVAILABLE=1 WITH_SANITIZE_MEMORY=OFF WITH_SANITIZE_MEMORY_AVAILABLE=1 WITH_SANITIZE_THREAD=OFF WITH_SANITIZE_THREAD_AVAILABLE=1 WITH_SERVER=ON WITH_SERVER_CHANNELS=ON WITH_SERVER_INTERFACE=ON WITH_SHADOW=ON WITH_SHADOW_MAC=ON WITH_SHADOW_X11=ON WITH_SMARTCARD_INSPECT=OFF WITH_SOXR=ON WITH_SSE2=ON WITH_SWSCALE=OFF WITH_THIRD_PARTY=OFF WITH_VAAPI=ON WITH_VALGRIND_MEMCHECK=OFF WITH_VALGRIND_MEMCHECK_AVAILABLE=1 WITH_WAYLAND=ON WITH_WINPR_TOOLS=ON WITH_X11=ON WITH_X264=OFF WITH_XCURSOR=ON WITH_XDAMAGE=ON WITH_XEXT=ON WITH_XFIXES=ON WITH_XI=ON WITH_XINERAMA=ON WITH_XKBFILE=ON WITH_XRANDR=ON WITH_XRENDER=ON WITH_XSHM=ON WITH_XTEST=ON WITH_XV=ON WITH_ZLIB=ON
Build type:          RelWithDebInfo
CFLAGS:              -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -g -fPIE -pie -DNDEBUG -fPIC -Wall -Wno-unused-result -Wno-unused-but-set-variable -Wno-deprecated-declarations -fvisibility=hidden -Wimplicit-function-declaration -Wredundant-decls -g -fno-omit-frame-pointer -DWINPR_DLL
Compiler:            GNU, 11.2.1
Target architecture: x64
```
* OS version connecting to Windows 2112
* If available the log output from a run with 
* If you built it yourself add some notes which tag/commit/branch you have used, also your cmake parameters and
  compiler can help

**Environment (please complete the following information):**
 - OS: OpenSUSE Tumbleweed
 - Version/Distribution: 20211101
 - Architecture: amd64

I have the same issue with both version 2.4.1 and 2.5.0, Wayland only. Sometimes after minutes, sometimes after a day or so. 

**Application details**

- FreeRDP version 2.4.1 and 2.5.0 
- Command line: /usr/bin/wlfreerdp +clipboard /w:3310 /h:1440 /bpp:32 -decorations /v:localhost:3393 /log-level:INFO -sec-nla /audio-mode:1 +offscreen-cache
- Build details:
```Build type      Gentoo
Install path    /usr
Compiler flags:
C               -O2 -pipe -march=native -fPIC -Wall -Wno-unused-result -Wno-unused-but-set-variable -Wno-deprecated-declarations -fvisibility=hidden -Wimplicit-function-declaration -Wredundant-decls -g -fno-omit-frame-pointer -DWINPR_DLL
C++             -O2 -pipe -march=native -Wno-deprecated-declarations -g
Linker flags:
Executable      -Wl,-O1 -Wl,--as-needed
Module          -Wl,-O1 -Wl,--as-needed
Shared          -Wl,-O1 -Wl,--as-needed
...
 * QA Notice: Package triggers severe warnings which indicate that it
 *            may exhibit random runtime failures.
 * /var/tmp/portage/net-misc/freerdp-2.5.0/work/freerdp-2.5.0/winpr/libwinpr/asn1/asn1.c:102:40: warning: array subscript 1 is outside array bounds of ‘struct ASN1encoding_s[1]’ [-Warray-bounds]
 * /var/tmp/portage/net-misc/freerdp-2.5.0/work/freerdp-2.5.0/winpr/libwinpr/asn1/asn1.c:97:40: warning: array subscript 1 is outside array bounds of ‘struct ASN1encoding_s[1]’ [-Warray-bounds]

 * Please do not file a Gentoo bug and instead report the above QA
 * issues directly to the upstream developers of this software.
 * Homepage: http://www.freerdp.com/
```
- Stacktrace:
```[18:23:31:461] [12548:12548] [WARN][com.freerdp.client.wayland.pointer] - wlf_Pointer_SetPosition not implemented
malloc(): unaligned tcache chunk detected
[19:33:16:424] [12548:12548] [ERROR][com.freerdp.utils] - Caught signal 'Aborted' [6]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 0: /usr/bin/../lib64/libwinpr2.so.2(winpr_backtrace+0x50) [0x7f9794ca2190]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 1: /usr/bin/../lib64/libwinpr2.so.2(winpr_log_backtrace_ex+0x2f) [0x7f9794ca232f]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 2: /usr/bin/../lib64/libfreerdp2.so.2(+0x304a0) [0x7f9794daa4a0]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 3: /lib64/libc.so.6(+0x38b80) [0x7f9794a95b80]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 4: /lib64/libc.so.6(gsignal+0x141) [0x7f9794a95b01]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 5: /lib64/libc.so.6(abort+0x112) [0x7f9794a7f536]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 6: /lib64/libc.so.6(+0x7b817) [0x7f9794ad8817]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 7: /lib64/libc.so.6(+0x836ba) [0x7f9794ae06ba]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 8: /lib64/libc.so.6(__libc_malloc+0x30c) [0x7f9794ae54fc]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 9: /usr/bin/../lib64/libfreerdp2.so.2(+0xb0408) [0x7f9794e2a408]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 10: /usr/bin/../lib64/libfreerdp2.so.2(+0xa66f6) [0x7f9794e206f6]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 11: /usr/bin/../lib64/libfreerdp2.so.2(+0xa6bce) [0x7f9794e20bce]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 12: /usr/bin/../lib64/libfreerdp2.so.2(+0x9d71a) [0x7f9794e1771a]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 13: /usr/bin/../lib64/libfreerdp2.so.2(+0xa0967) [0x7f9794e1a967]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 14: /usr/bin/../lib64/libfreerdp2.so.2(+0xaabfc) [0x7f9794e24bfc]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 15: /usr/bin/../lib64/libfreerdp2.so.2(+0xa10f3) [0x7f9794e1b0f3]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 16: /usr/bin/../lib64/libfreerdp2.so.2(freerdp_check_fds+0x3f) [0x7f9794e01c2f]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 17: /usr/bin/../lib64/libfreerdp2.so.2(freerdp_check_event_handles+0x13) [0x7f9794e02d33]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 18: /usr/bin/wlfreerdp(+0x391e) [0x56489d11c91e]
[19:33:16:428] [12548:12548] [ERROR][com.freerdp.utils] - 19: /lib64/libc.so.6(__libc_start_main+0xcd) [0x7f9794a807fd]
Aborted
```

Based on the QA message from the build and the stack trace, it looks like the issue is with asn1.c

--- a/include/freerdp/update.h
+++ b/include/freerdp/update.h
@@ -66,7 +66,6 @@ typedef struct _BITMAP_DATA BITMAP_DATA;
 
 struct _BITMAP_UPDATE
 {
-	UINT32 count;
 	UINT32 number;
 	BITMAP_DATA* rectangles;
 	BOOL skipCompression;

--- a/libfreerdp/core/update.c
+++ b/libfreerdp/core/update.c
@@ -215,20 +215,10 @@ BITMAP_UPDATE* update_read_bitmap_update(rdpUpdate* update, wStream* s)
 	Stream_Read_UINT16(s, bitmapUpdate->number); /* numberRectangles (2 bytes) */
 	WLog_Print(up->log, WLOG_TRACE, "BitmapUpdate: %" PRIu32 "", bitmapUpdate->number);
 
-	if (bitmapUpdate->number > bitmapUpdate->count)
-	{
-		UINT32 count = bitmapUpdate->number * 2;
-		BITMAP_DATA* newdata =
-		    (BITMAP_DATA*)realloc(bitmapUpdate->rectangles, sizeof(BITMAP_DATA) * count);
+	bitmapUpdate->rectangles = (BITMAP_DATA*)calloc(bitmapUpdate->number, sizeof(BITMAP_DATA));
 
-		if (!newdata)
-			goto fail;
-
-		bitmapUpdate->rectangles = newdata;
-		ZeroMemory(&bitmapUpdate->rectangles[bitmapUpdate->count],
-		           sizeof(BITMAP_DATA) * (count - bitmapUpdate->count));
-		bitmapUpdate->count = count;
-	}
+	if (!bitmapUpdate->rectangles)
+		goto fail;
 
 	/* rectangles */
 	for (i = 0; i < bitmapUpdate->number; i++)

--- a/server/shadow/shadow_client.c
+++ b/server/shadow/shadow_client.c
@@ -1535,7 +1535,7 @@ static BOOL shadow_client_send_bitmap_update(rdpShadowClient* client, BYTE* pSrc
 	cols = (nWidth / 64) + ((nWidth % 64) ? 1 : 0);
 	k = 0;
 	totalBitmapSize = 0;
-	bitmapUpdate.count = bitmapUpdate.number = rows * cols;
+	bitmapUpdate.number = rows * cols;
 
 	if (!(bitmapData = (BITMAP_DATA*)calloc(bitmapUpdate.number, sizeof(BITMAP_DATA))))
 		return FALSE;
@@ -1613,8 +1613,8 @@ static BOOL shadow_client_send_bitmap_update(rdpShadowClient* client, BYTE* pSrc
 		}
 	}
 
-	bitmapUpdate.count = bitmapUpdate.number = k;
-	updateSizeEstimate = totalBitmapSize + (k * bitmapUpdate.count) + 16;
+	bitmapUpdate.number = k;
+	updateSizeEstimate = totalBitmapSize + (k * bitmapUpdate.number) + 16;
 
 	if (updateSizeEstimate > maxUpdateSize)
 	{
@@ -1649,7 +1649,7 @@ static BOOL shadow_client_send_bitmap_update(rdpShadowClient* client, BYTE* pSrc
 
 			if ((newUpdateSize >= maxUpdateSize) || (i + 1) >= k)
 			{
-				bitmapUpdate.count = bitmapUpdate.number = j;
+				bitmapUpdate.number = j;
 				IFCALLRET(update->BitmapUpdate, ret, context, &bitmapUpdate);
 
 				if (!ret)

--- a/uwac/libuwac/uwac-priv.h
+++ b/uwac/libuwac/uwac-priv.h
@@ -255,6 +255,14 @@ struct uwac_window
 	int pointer_current_cursor;
 };
 
+/**@brief data to pass to wl_buffer release listener */
+struct uwac_buffer_release_data
+{
+	UwacWindow* window;
+	int bufferIdx;
+};
+typedef struct uwac_buffer_release_data UwacBufferReleaseData;
+
 /* in uwa-display.c */
 UwacEvent* UwacDisplayNewEvent(UwacDisplay* d, int type);
 int UwacDisplayWatchFd(UwacDisplay* display, int fd, uint32_t events, UwacTask* task);

--- a/uwac/libuwac/uwac-window.c
+++ b/uwac/libuwac/uwac-window.c
@@ -48,7 +48,8 @@ static int bppFromShmFormat(enum wl_shm_format format)
 
 static void buffer_release(void* data, struct wl_buffer* buffer)
 {
-	UwacBuffer* uwacBuffer = (UwacBuffer*)data;
+	UwacBufferReleaseData* releaseData = data;
+	UwacBuffer* uwacBuffer = &releaseData->window->buffers[releaseData->bufferIdx];
 	uwacBuffer->used = false;
 }
 
@@ -66,7 +67,10 @@ static void UwacWindowDestroyBuffers(UwacWindow* w)
 #else
 		region16_uninit(&buffer->damage);
 #endif
+		UwacBufferReleaseData* releaseData =
+		    (UwacBufferReleaseData*)wl_buffer_get_user_data(buffer->wayland_buffer);
 		wl_buffer_destroy(buffer->wayland_buffer);
+		free(releaseData);
 		munmap(buffer->data, buffer->size);
 	}
 
@@ -344,7 +348,8 @@ int UwacWindowShmAllocBuffers(UwacWindow* w, int nbuffers, int allocSize, uint32
 
 	for (i = 0; i < nbuffers; i++)
 	{
-		UwacBuffer* buffer = &w->buffers[w->nbuffers + i];
+		int bufferIdx = w->nbuffers + i;
+		UwacBuffer* buffer = &w->buffers[bufferIdx];
 #ifdef HAVE_PIXMAN_REGION
 		pixman_region32_init(&buffer->damage);
 #else
@@ -354,7 +359,10 @@ int UwacWindowShmAllocBuffers(UwacWindow* w, int nbuffers, int allocSize, uint32
 		buffer->size = allocSize;
 		buffer->wayland_buffer =
 		    wl_shm_pool_create_buffer(pool, allocSize * i, width, height, w->stride, format);
-		wl_buffer_add_listener(buffer->wayland_buffer, &buffer_listener, buffer);
+		UwacBufferReleaseData* listener_data = xmalloc(sizeof(UwacBufferReleaseData));
+		listener_data->window = w;
+		listener_data->bufferIdx = bufferIdx;
+		wl_buffer_add_listener(buffer->wayland_buffer, &buffer_listener, listener_data);
 	}
 
 	wl_shm_pool_destroy(pool);

